File: ios/WalletCore.m
#import <React/RCTBridgeModule.h>
#import <Foundation/Foundation.h>
#import <GoCore/GoCore.h>
#include "RustCore.h"

@interface RCT_EXTERN_MODULE(WalletCore, NSObject)

// explanation about threading here: https://stackoverflow.com/a/50775641/3060739
- (dispatch_queue_t)methodQueue
{
  return dispatch_get_main_queue();
}

+ (BOOL)requiresMainQueueSetup
{
  return YES;
}

RCT_EXTERN_METHOD(ironfishCreateAccount:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(ironfishCreateAccountFromPrivateKey:(NSString*)pk resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(ironfishSignTransaction:(NSString*)rawTx pk:(NSString*)pk resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(ironfishInitSaplingParams:(NSString*)mintParamsPath spendParamsPath:(NSString*)spendParamsPath outputParamsPath:(NSString*)outputParamsPath resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(ironfishIsValidAddress:(NSString*)address resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(aleoCreateAccountFromSeed:(NSString*)seed resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(aleoCreateAccountFromPrivateKey:(NSString*)pk resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(aleoIsValidAddress:(NSString*)address resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(aleoDeserializeCreditsRecord:(NSString*)recordStr resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(aleoSignMessage:(NSString*)pk message:(NSString*)message resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(aleoGenerateProverFiles:(NSString*)dir rpc_url:(NSString*)rpc_url network:(NSString*)network program_id:(NSString*)program_id function_name:(NSString*)function_name prover_file_path:(NSString*)prover_file_path verifier_file_path:(NSString*)verifier_file_path resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(aleoExecuteProgram:(NSString*)dir rpc_url:(NSString*)rpc_url network:(NSString*)network pk:(NSString*)pk program_id:(NSString*)program_id function_name:(NSString*)function_name inputs:(NSString*)inputs fee_record:(NSString*)fee_record base_fee:(NSString*)base_fee priority_fee:(NSString*)priority_fee prover_file_path:(NSString*)prover_file_path verifier_file_path:(NSString*)verifier_file_path fee_prover_file_path:(NSString*)fee_prover_file_path fee_verifier_file_path:(NSString*)fee_verifier_file_path resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(aleoDecryptRecord:(NSString*)ciphertext vk:(NSString*)vk resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(aleoParseRecord:(NSString*)plaintext resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(aleoHashBhp256:(NSString*)str resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)

RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(ironfishCreateAccountFromPrivateKeySync:(NSString*)pk) {
  const char* resPtr = ironfish_create_account_from_pk(NULL, [pk UTF8String]);
  NSString *res = [[NSString alloc] initWithUTF8String:resPtr];
  core_destroy_string(resPtr);
  return res;
}

RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(ironfishIsValidAddressSync:(NSString*)address) {
  const char* resPtr = ironfish_is_valid_public_address(NULL, [address UTF8String]);
  NSString *res = [[NSString alloc] initWithUTF8String:resPtr];
  core_destroy_string(resPtr);
  return res;
}

RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(aleoCreateAccountFromSeedSync:(NSString*)seed) {
  const char* resPtr = aleo_create_account_from_seed(NULL, [seed UTF8String]);
  NSString *res = [[NSString alloc] initWithUTF8String:resPtr];
  core_destroy_string(resPtr);
  return res;
}

RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(aleoCreateAccountFromPrivateKeySync:(NSString*)pk) {
  const char* resPtr = aleo_create_account_from_pk(NULL, [pk UTF8String]);
  NSString *res = [[NSString alloc] initWithUTF8String:resPtr];
  core_destroy_string(resPtr);
  return res;
}

RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(aleoIsValidAddressSync:(NSString*)address) {
  const char* resPtr = aleo_is_valid_public_address(NULL, [address UTF8String]);
  NSString *res = [[NSString alloc] initWithUTF8String:resPtr];
  core_destroy_string(resPtr);
  return res;
}

RCT_EXTERN_METHOD(spaceMeshAddressFromPubKey:(NSString*)pub resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)

RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(spaceMeshAddressFromPubKeySync:(NSString*)pub) {
  return FoxSpaceMeshAddressFromPubKey(pub);
}
RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(spaceMeshStringToAddressSync:(NSString*)address) {
  return FoxSpaceMeshStringToAddress(address);
}

RCT_EXTERN_METHOD(spaceMeshCreateTransaction:(NSString*)pk to:(NSString*)to amount:(NSString*)amount nonce:(NSString*)nonce gasPrice:(NSString*)gasPrice genesisID:(NSString*)genesisID resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(spaceMeshSelfSpawnTx:(NSString*)pk nonce:(NSString*)nonce gasPrice:(NSString*)gasPrice genesisID:(NSString*)genesisID resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(polkadotScrypt:(NSDictionary*)params resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject) {
  NSString *passphrase = [params[@"passphrase"] stringValue];
  NSString *salt = [params[@"salt"] stringValue];
  NSInteger log2_n = [params[@"log2_n"] unsignedIntValue];
  NSInteger r = [params[@"r"] unsignedIntValue];
  NSInteger p = [params[@"p"] unsignedIntValue];

  const char* resPtr = polkadot_scrypt(NULL, [passphrase UTF8String], [salt UTF8String], log2_n, r, p);
  NSString *res = [[NSString alloc] initWithUTF8String:resPtr];
  resolve(res);
}

@end

File: ios/WalletCore.swift
import Foundation

func handle_error<T, U>(
  resolve: RCTPromiseResolveBlock,
  reject: RCTPromiseRejectBlock,
  get_result: (UnsafeMutablePointer<ExternError>) -> T,
  success: (T) -> U
) -> Void {
  var err = ExternError()
  let err_ptr: UnsafeMutablePointer<ExternError> = UnsafeMutablePointer(&err)
  let res = get_result(err_ptr)
  if err_ptr.pointee.code == 0 {
    resolve(success(res))
  } else {
    let val = String(cString: err_ptr.pointee.message)
    core_destroy_string(err_ptr.pointee.message)
    reject(String(describing: err_ptr.pointee.code), val, nil)
  }
}

@objc(WalletCore)
class WalletCore: NSObject {

  public static func requiresMainQueueSetup() -> Bool {
    return true;
  }

  @objc func ironfishCreateAccount(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    handle_error(
      resolve: resolve,
      reject: reject,
      get_result: { ironfish_create_account($0) },
      success: { (res: Optional<UnsafePointer<CChar>>) -> String in
        let val = String(cString: res!)
        core_destroy_string(res!)
        return val
    })
  }

  @objc func ironfishCreateAccountFromPrivateKey(_ pk: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    handle_error(
      resolve: resolve,
      reject: reject,
      get_result: { ironfish_create_account_from_pk($0, pk) },
      success: { (res: Optional<UnsafePointer<CChar>>) -> String in
        let val = String(cString: res!)
        core_destroy_string(res!)
        return val
    })
  }

  @objc func ironfishSignTransaction(_ rawTx: String, pk: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { ironfish_sign_transaction($0, rawTx, pk) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func ironfishInitSaplingParams(_ mintParamsPath: String, spendParamsPath: String, outputParamsPath: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { ironfish_init_sapling_params($0, mintParamsPath, spendParamsPath, outputParamsPath) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func ironfishIsValidAddress(_ address: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { ironfish_is_valid_public_address($0, address) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func aleoCreateAccountFromSeed(_ seed: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { aleo_create_account_from_seed($0, seed) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func aleoCreateAccountFromPrivateKey(_ pk: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { aleo_create_account_from_pk($0, pk) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func aleoIsValidAddress(_ address: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { aleo_is_valid_public_address($0, address) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func aleoDeserializeCreditsRecord(_ recordStr: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { aleo_deserialize_credits_record($0, recordStr) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func aleoSignMessage(_ pk: String, message: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { aleo_sign_message($0, pk, message) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func aleoGenerateProverFiles(_ dir: String, rpc_url: String, network: String, program_id: String, function_name: String, prover_file_path: String, verifier_file_path: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { aleo_generate_prover_files($0, dir, rpc_url, network, program_id, function_name, prover_file_path, verifier_file_path) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func aleoExecuteProgram(_ dir: String, rpc_url: String, network: String, pk: String, program_id: String, function_name: String, inputs: String, fee_record: String, base_fee: String, priority_fee: String, prover_file_path: String, verifier_file_path: String, fee_prover_file_path: String, fee_verifier_file_path: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { aleo_execute_program($0, dir, rpc_url, network, pk, program_id, function_name, inputs, fee_record, base_fee, priority_fee, prover_file_path, verifier_file_path, fee_prover_file_path, fee_verifier_file_path) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func aleoDecryptRecord(_ ciphertext: String, vk: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { aleo_decrypt_record($0, ciphertext, vk) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func aleoParseRecord(_ plaintext: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { aleo_parse_record($0, plaintext) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func aleoHashBhp256(_ str: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
   handle_error(
     resolve: resolve,
     reject: reject,
     get_result: { aleo_hash_bhp256($0, str) },
     success: { (res: Optional<UnsafePointer<CChar>>) -> String in
       let val = String(cString: res!)
       core_destroy_string(res!)
       return val
   })
  }

  @objc func spaceMeshAddressFromPubKey(_ pub: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
    resolve(FoxSpaceMeshAddressFromPubKey(pub))
  }

  @objc func spaceMeshCreateTransaction(_ pk: String, to: String, amount: String, nonce: String, gasPrice: String, genesisID: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
    resolve(FoxSpaceMeshCreateTransaction(pk, to, amount, nonce, gasPrice, genesisID))
  }

  @objc func spaceMeshSelfSpawnTx(_ pk: String, nonce: String, gasPrice: String, genesisID: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
    resolve(FoxSpaceMeshSelfSpawnTx(pk, nonce, gasPrice, genesisID))
  }
}

